eta/core/datasets.py

```diff
@@ -1372,18 +1375,336 @@ def transform(self, src):
+                2 - will square the values
+                ...
+                Check Balancer._solution_score for more details.
+        '''
+        self.attr_name = attribute_name
+        self.object_label = object_label
+        self.median_point = median_point
+        self.negative_power = negative_power
 
     def transform(self, src):
-        # @TODO implement Balancing!!
+        '''
+        Modify the BuilderDataset records by removing records until the
+        target attribute is ~roughly~ balanced for each value.
+
+        Args:
+            src (BuilderDataset): the dataset builder
+        '''
+
+        # STEP 1: Get attribute value(s) for every record
+        helper_list = self._to_helper_list(src.records)
+        if not len(helper_list):
+            return
+
+        # STEP 2: determine target number to remove of each attribute value
+        target_remove = self._get_target_remove(helper_list)
+
+        # STEP 3: find the records to keep
+        keep_idxs = self._get_keep_idxs(target_remove, helper_list)
+
+        # STEP 4: modify the list of records
         old_records = src.records
         src.clear()
-        pass
+        for ki in keep_idxs:
+            src.add(old_records[helper_list[ki][0]])
+
+    def _to_helper_list(self, records):
+        '''
+
+        records: list of BuilderDataRecord's
+        return: a list of tuples with two entries:
+
+                [
+                    (record_id, list_of_values),
+                    (record_id, list_of_values),
+                    (record_id, list_of_values),
+                    ...
+                ]
+
+            record_id: integer ID of the corresponding old_record
+            list_of_values: list of attribute values for the attribute to be
+                            balanced, one per unique object, if using objects.
+                            For example: ['red', 'red', 'green'] would imply
+                            three objects with the 'color' attribute in this
+                            record.
+        '''
+
+        if not len(records):
+            return []
+
+        elif type(records[0]) == BuilderImageRecord:
+            if self.object_label:
+                return self._to_helper_list_image_objects(records)
+            else:
+                return self._to_helper_list_image(records)
+
+        elif type(records[0]) == BuilderVideoRecord:
+            if self.object_label:
+                return self._to_helper_list_video_objects(records)
+            else:
+                return self._to_helper_list_video(records)
+
+        raise DatasetTransformerError(
+            "Unknown record type: {}".format(type(records[0])))
+
+    def _to_helper_list_image(self, records):
+        '''Balancer._to_helper_list for image attributes'''
+        helper_list = []
+
+        for i, record in enumerate(records):
+            labels = record.get_labels()
+
+            for attr in labels.attrs:
```

Comment:
Maybe we should be keeping only `CategoricalAttribute`s and `BooleanAttribute`s here, (i.e. not `NumericAttribute`s).

Reply:
I've been assuming categorical...but would there be an issue for it to be boolean or numeric? I don't see the problem. unless perhaps numeric is a continuous space...